Added save/cancel buttons to deck options

bhatushar opened the pull request:
## Purpose / Description
Providing user Save and Cancel buttons to explicitly decide whether to save changes in deck options or discard them. Originally, some changes were not saved until the user went back to the deck screen. In addition to that, some of the changes may be partially committed if the app is closed immediately after changing a setting.

## Fixes
Fixes #8180 

## Approach
Added a 'Save' button which triggers a database write. This eliminates any ambiguity on the user's end. In addition, a 'Cancel' button is also added to undo all temporary changes. Pressing the back key after making changes or changing options group opens a confirmation dialog.
Changing settings in Group Management immediately save all other changes made by the user. A note is added to inform the user of the same.

## How Has This Been Tested?
Emulated on API 29, Android 10.0. Worked as expected.

## UI Changes
| Deck options screen | Confirmation on back key press |
|:-------------------------:|:-------------------------:|
| ![main-screen](https://user-images.githubusercontent.com/21220253/111863341-1588b780-8981-11eb-9b73-bae4c5799610.png) | ![back-key-confirm](https://user-images.githubusercontent.com/21220253/111863391-508aeb00-8981-11eb-9330-4954cee7f648.PNG) |

| Confirmation on changing Options Group | Added note in Group Management screen |
|:-------------------------:|:-------------------------:|
| ![options-group-change](https://user-images.githubusercontent.com/21220253/111863411-68fb0580-8981-11eb-87fd-0ab1ffb0377f.PNG) | ![group-management-screen](https://user-images.githubusercontent.com/21220253/111863414-744e3100-8981-11eb-99a8-0bba6816d25b.PNG) |


## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 reviewed (commented):
Thanks so much! This looks very clean from first inspections

This is a lot of code to review, please ping me if I (or another maintainer) haven't been around to it by Monday.

* Could you save the refactoring in a7e2aaf8766f1898b1e28a23e34a9cefac399b29 for a later  pull request (unless there are functional changes)? This is already a complex review, and will add time to the review.
* Is it possible to use "Back" button on the Action Bar, and add `Save` there, rather than a preference? (Maybe not as we're using `AppCompatPreferenceActivity`).
* There are some lint failures related to duplicate strings `gradelw lintRelease` should explain these

bhatushar commented:
- I had a feeling these changes will be hard to review that's why I left the refactoring for last. Those changes are so that future developers have an easier time navigating the code. I'll roll back to the previous commit for now. After the review, should I redo those changes in the same PR or create a new one?
- At the moment, pressing the Back button in Action Bar, opens a dialog that asks for user confirmation. Would you rather it saves by default?
- I looked into that and found two duplicates. I'll fix that.

4shutosh commented:
@bhatushar I suggest you never do a force push as it tampers the commit history because it overwrites history on GitHub and makes the incremental changes harder to review. 

NOT a good practice.

bhatushar commented:
@4shutosh I know. I had to revert a commit, but accidentally goofed up. To fix up my mess, I force pushed this branch to an older commit.

This was the only way I could think of fixing it while keeping things relatively clean.

david-allison-1 commented:
> @bhatushar I suggest you never do a force push as it tampers the commit history because it overwrites history on GitHub and makes the incremental changes harder to review.
> 
> NOT a good practice.

It's common to rebase PRs on this repository (note: NOT any public branches). 

We value a clean `git log` for the project. It's exceptionally rare for anyone but the PR author to push to their PR (so they're de-facto private) and it means that we can ensure that commits are properly sized in a review before we rebase onto master.

Not rebasing would either mean that when we merge to master:

* we `squash`, removing granular commit data
* we `rebase`, adding WIP commits into the history, leading to a bloated and confusing log

bhatushar commented:
> This is a lot of code to review, please ping me if I (or another maintainer) haven't been around to it by Monday.

@david-allison-1 @Arthur-Milchior @mikehardy Is anyone reviewing this PR?

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -104,6 +120,14 @@ public void saveResults(SaveStatus status) {
+    /**
+     * @return Current active instance of DeckOptions
+     */
+    public static DeckOptions getCurrentInstance() {
```
We shouldn't be using static variables unless we have to. What are the other option here?

bhatushar replied:
Resolved in 3d41c495

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -83,27 +83,132 @@
+     * Keeps track of the most recent instance of {@link DeckOptions}, which is used by
+     * {@link com.ichi2.preferences.SaveDeckOptionsPreference} to send save signals.
+     *
+     * FIXME: Modify to support multiple instances
```
Flagging this - this causes a context/memory leak.

bhatushar replied:
Resolved in 3d41c49

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -83,27 +83,132 @@
+         * Set of XML preference keys that are modified. Used to discard changes.
+         * Keys are added inside {@link Editor#commit()} using data of {@link Editor#mUpdate} variable.
+         */
+        HashSet<String> changes = new HashSet<>();
```
This is unused

bhatushar replied:
This was being used in `commit()` and `resetValues()`. Now it's used inside `commit()` and `writeChanges()`.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -83,27 +83,132 @@
+     * Boolean variables are used to keep track of these preferences inside Editor.commit().
+     * These settings are applied in DeckOptions.updatePostSaveSettings().
+     */
+    private boolean updateOrder = false;
```
note: naming (`m` prefix), but these should go in the new design

bhatushar replied:
Resolved in 534a72b3 and efd5cd59.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -303,8 +555,48 @@ public boolean commit() {
                                 break;
                             case "deckConf": {
```
Design: might be best to separate changes which will save, and ones that won't

bhatushar replied:
What do you propose?

I can think of two solutions:

1. Create a separate setting for Group Management. I suspect it could lead to a lot of duplicate code.
![group-management](https://user-images.githubusercontent.com/21220253/112657539-898ef800-8e78-11eb-888b-630a08ebeb3b.PNG)

2. Pop-up an alert whenever the user opens the Group Management section. If possible, it might be the simplest solution.

david-allison-1 replied:
I meant from a code design perspective, not from a visual perspective. 

We do need to handle how "group management" will force people to save, and I think this can be done via a dialog.

david-allison-1 commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -184,6 +184,7 @@
 
     <!-- Deck Options -->
     <string name="deck_options_corrupt">Failed to process deck options: %s</string>
+    <string name="deck_options_group_changes" comment="XXX is dynamically replaced with options group name">Do you want to save changes for XXX?</string>
```
```suggestion
    <string name="deck_options_group_changes" comment="XXX is dynamically replaced with options group name">Do you want to save changes for 'XXX'?</string>
```

bhatushar replied:
Resolved in 78d9a615.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -303,8 +555,48 @@ public boolean commit() {
+                                            .getDefault(getDeckOptions())
+                                            // confirmationText = "Do you want to save changes for {Options group name}?"
+                                            .content(confirmationText)
+                                            .positiveText(getResources().getString(R.string.yes))
```
Use "Save"/"Cancel" here:

> Don’t use action text that fails to indicate what the selection will do. “Cancel” and “Delete” better indicate what will occur in this dialog.

https://material.io/components/dialogs#alert-dialog

bhatushar replied:
I think "Delete" or "Discard" might be better suited here than "Cancel" because the prompt asks "Do you want to save changes for XXX?". "Cancel" could be inferred as not doing anything. 

What do you think I should use?

david-allison-1 replied:
Ideally we don't want to be in a situation where the user has to make a "save/discard" choice without inspecting the settings. Can we design the dialog in another way?

At worst case, we want save/discard/cancel

bhatushar replied:
Is this good?
![new-dialog](https://user-images.githubusercontent.com/21220253/112638028-cea92f00-8e64-11eb-9d9d-e61bb406a7eb.PNG)

david-allison-1 replied:
'Discard' makes more sense

david-allison-1 reviewed (changes requested):
Thanks!! And sorry for the delay - large reviews take a large time block

**UI**

* Move the Save/Cancel to the App Bar as a save button, keep the original "back" icon]
* The warning in "Group management" that settings will be saved immediately is not prominent enough

----
**Design**

* `commit()` is called each time a preference changes
   * This should be documented

So, currently, we're modifying `mOptions`, and reverting this change if `cancel` is pressed. This is a dangerous operation.

Instead, we should be generating a list of changed state, as a list of changes, and applying them when `save` is pressed - this means that there's no chance of accidentally forgetting to undo data. It'll also avoid the crash below

After this, `resetValues()` will be able to be removed, and there should be a lot less code changed.

----
**Crash**

* Options
* Reminder
* Enable
* Back
* Cancel

```
2021-03-24 08:15:13.817 17468-17468/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 17468
    com.ichi2.utils.JSONException: org.json.JSONException: No value for reminder
        at com.ichi2.utils.JSONObject.get(JSONObject.java:240)
        at com.ichi2.anki.DeckOptions$DeckPreferenceHack.resetValues(DeckOptions.java:408)
        at com.ichi2.anki.DeckOptions.closeWithResult(DeckOptions.java:1022)
        at com.ichi2.anki.DeckOptions.saveResults(DeckOptions.java:145)
        at com.ichi2.preferences.SaveDeckOptionsPreference.lambda$onCreateView$1(SaveDeckOptionsPreference.java:90)
        at com.ichi2.preferences.-$$Lambda$SaveDeckOptionsPreference$uylKe53tmTfv-vP4YzedK6D9bmE.onClick(Unknown Source:0)
        at android.view.View.performClick(View.java:7448)
        at android.view.View.performClickInternal(View.java:7425)
        at android.view.View.access$3600(View.java:810)
        at android.view.View$PerformClick.run(View.java:28305)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7660)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: org.json.JSONException: No value for reminder
        at org.json.JSONObject.get(JSONObject.java:400)
        at com.ichi2.utils.JSONObject.get(JSONObject.java:238)
        at com.ichi2.anki.DeckOptions$DeckPreferenceHack.resetValues(DeckOptions.java:408) 
        at com.ichi2.anki.DeckOptions.closeWithResult(DeckOptions.java:1022) 
        at com.ichi2.anki.DeckOptions.saveResults(DeckOptions.java:145) 
        at com.ichi2.preferences.SaveDeckOptionsPreference.lambda$onCreateView$1(SaveDeckOptionsPreference.java:90) 
        at com.ichi2.preferences.-$$Lambda$SaveDeckOptionsPreference$uylKe53tmTfv-vP4YzedK6D9bmE.onClick(Unknown Source:0) 
        at android.view.View.performClick(View.java:7448) 
        at android.view.View.performClickInternal(View.java:7425) 
        at android.view.View.access$3600(View.java:810) 
        at android.view.View$PerformClick.run(View.java:28305) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7660) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
```

4shutosh commented on AnkiDroid/src/main/java/com/ichi2/preferences/SaveDeckOptionsPreference.java:
```diff
@@ -0,0 +1,98 @@
+/****************************************************************************************
+ * Copyright (c) 2021 Tushar Bhatt <tbhatt321@gmail.com>                                *
+ *                                                                                      *
```
I don't think these comments are required

bhatushar replied:
It's in accordance to their code style. Take a look at this: https://github.com/ankidroid/Anki-Android/issues/8211

4shutosh replied:
Thank you.

bhatushar commented:
@david-allison-1 Just clarifying a few things.

About the position of the save and cancel buttons, are you proposing something like this? If so, I'm not sure if that's possible but I'll look into it and let you know.
![screen](https://user-images.githubusercontent.com/21220253/112304860-0331b500-8cc4-11eb-9a6e-bf437a14685e.jpg)


Also, about that crash! I'm switching to your proposed method of applying changes when saving, instead of undoing. So that crash won't be a problem but I thought it might be helpful to explain why that happened. I was creating a copy of `mOptions` inside `onCreate()` and I was accessing the "reminder" key of that copy inside `reset()`. However, `mOptions` doesn't always have a "reminder" key. It is added inside `commit()`.

david-allison-1 commented:
Thanks! Something along the lines of this (except without the second "reset" button): 

![image](https://user-images.githubusercontent.com/62114487/112306037-44f83680-8c97-11eb-9d09-b10c36b30591.png)

bhatushar commented:
@david-allison-1 Does this satisfy the requirements? It also solves the memory leak problem. The back button still asks for confirmation. Do you want me to disable that?

![save-option](https://user-images.githubusercontent.com/21220253/112644965-ffd92d80-8e6b-11eb-86b3-24ba605f9103.PNG)

david-allison-1 commented:
No, confirmation is great, this looks good!

bhatushar commented:
> Let me know when it's ready for re-review

Soon! All requests are resolved except for https://github.com/ankidroid/Anki-Android/pull/8285#discussion_r600269934. Let me just double-check everything before pushing.

bhatushar commented:
@david-allison-1 These changes are ready to be reviewed. I'm going to discuss a possible solution for https://github.com/ankidroid/Anki-Android/pull/8285#discussion_r600269934 in it's respective thread.

## Changelist
- Writing changes to DB on pressing save instead of undoing changes on cancel.
	- On preference change, copies of Deck and Options are modified instead of actual data.
	- On saving, changes in copies are applied to original data before writing to DB.
- Options data is not reloaded inside `cacheValues()`.
- Renamed variables in 534a72b3. These changes are only for better readability and don't affect the functionality.
- Implemented save option in the menu.
	- Removed Save/Cancel buttons from the bottom of the preferences.
	- Removed static variable which was causing memory leak.

## Proposals
- Reording cases in `commit()` for better readability.
- Optimizing `commit()` cases.
	- Track changes which are undone by user. This avoids redundant DB writes.
	- Use mChanges instead of mPreferenceChanged. They both achieve the same thing functionally.
	- Removing mPreferenceChanged could lead to removing listener event at the end of `commit()`.

Arthur-Milchior commented:
Would you please change the top level comment, so that it's clear what's going on. I must admit that reviewing 28 commit is extremely frightening, I fear I'll need to spend more than an hour to do it properly. Knowing why it must take so much would be really helpful. And maybe splitting in logical distinct PR too

bhatushar commented:
@Arthur-Milchior what details would you like me to include?


> And maybe splitting in logical distinct PR too

All the changes only accomplish one thing: Giving user the option to explicitly save changes. I admit these are a lot of changes. Even I wasn't expecting this much when I started working on it.

Bottom line is, all changes are functionally interdependent (expect for refactorimg ones) and can't be made into separate PR.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -217,6 +279,150 @@ private boolean parseTimerValue(DeckConfig options) {
+         * This function writes original value to every changed property of {@link #mOptions}, so that the activity
+         * shows the correct data when user re-opens these settings.
+         */
+        public void resetValues() {
```
I believe there is far simpler. A deck option is simply a JSON Dict. You can copy the dict, make the change in the copy, and save the copy at the end if the save are confirmed.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -685,6 +895,12 @@ protected void onCreate(Bundle icicle) {
+        // Create backups
+        mDeckBackup = mDeck.deepClone();
+        mOptions = mCol.getDecks().confForDid(mDeck.getLong("id"));
+        mOptionsBackup = new DeckConfig(mOptions.toString()); // deepClone() not available for DeckConfig type
```
You can simply add `deepClone` on it. I just did it in #8350

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -129,12 +146,57 @@ public static DeckOptions getCurrentInstance() {
+
+        if (updateReminderEnabled || updateReminderTime) {
+            // Update reminder settings (key: reminderEnabled, reminderTime)
+            final AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
```
Since some part of this code is already in master branch, and is moved to another function, I would appreciate if you could create a short PR, exctracting this code to an external function. This will allow this commit to be smaller and simpler to check, since instead of moving blocks of code you'd be moving function calls.
Same remarks with other part of the codes that are moved

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -555,11 +555,19 @@ public boolean commit() {
                                     // Preferences are changed, ask whether to save/discard before proceeding
+                                    String confirmationText = getResources().getString(R.string.deck_options_group_changes);
+                                    try {
+                                        confirmationText = confirmationText.replace("XXX", mOptions.getString("name"));
```
You can use 
`mOptions.optString("name", "???")` instead of catching an exception.

Arthur-Milchior commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -105,6 +105,7 @@
 
     <string name="dialog_cancel">Cancel</string>
     <string name="dialog_ok">OK</string>
+    <string name="dialog_yes">Yes</string>
```
Yes is already in 01-core

files are not respected. Instead, we avoid duplicate to ensure that translators don't have to duplicate work uselessly

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -555,11 +555,19 @@ public boolean commit() {
+
                                     DiscardChangesDialog
                                             .getDefault(getDeckOptions())
-                                            .content("Do you want to save changes for " + mOptions.getString("name") + "?")
```
Thanks, I never knew that we still have some strings directly in text, it should not have been the case. It should be a PR by itself. It would be simpler to review, merge, and the current PR would be smaller.

Arthur-Milchior commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -184,6 +185,7 @@
 
     <!-- Deck Options -->
     <string name="deck_options_corrupt">Failed to process deck options: %s</string>
+    <string name="deck_options_group_changes" comment="XXX is dynamically replaced with options group name">Do you want to save changes for XXX?</string>
```
Look at other strings. We don't use "XXX" and replacement. We use "%s" to indicate there is a string variable. Then you call the string as argument in `getString`

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -90,12 +90,12 @@
     private BroadcastReceiver mUnmountReceiver = null;
     private DeckPreferenceHack mPref;
-    public enum SaveStatus {
+    public enum ChangeStatus {
```
Can you directly change the commit where you introduce this, instead of rewritting it? It makes change too complex. (same remark for most of "Minor refactoring" commit)

Arthur-Milchior commented on AnkiDroid/src/main/res/xml/deck_options.xml:
```diff
@@ -29,6 +29,10 @@
                 android:key="currentConf"
                 android:title="@string/deck_conf_current_group" />
 
+            <Preference
```
Can you please explain in the commit and here why you have `android:key="groupNotice"` ? I did search on it online, it seems to be related to notifications and it seems unrelated to current PR

bhatushar replied:
I used that key just in case someone needs to access it in the future. There's no particular reason why I chose that name in particular. At the moment, that key is not being used and the preference is only used to give a warning to the user about group management settings.

However, in @david-allison-1's review, he suggested that we use a different approach so this whole section might end up being removed.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/preferences/SaveDeckOptionsPreference.java:
```diff
@@ -65,6 +66,17 @@ protected View onCreateView(ViewGroup parent) {
         linearLayout.setOrientation(LinearLayout.HORIZONTAL);
         saveButton.setText(parent.getResources().getString(R.string.save));
         cancelButton.setText(parent.getResources().getString(R.string.cancel));
+        saveButton.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20);
```
Why this choiceof size?

I'd have expected we have something standard for buttons, but I can't find anything similar. Any opinion @mikehardy and @david-allison-1 ?

david-allison-1 replied:
This class will be going away and will move to the app bar. No preference.

Arthur-Milchior commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -369,7 +369,4 @@
     <string name="minus_sign" comment="Label for decrement button in IncrementerNumberRangePreference">-</string>
-
-    <!-- Deck Options -->
-    <string name="cancel">Cancel</string>
```
Can you refactor this onto the original commit where you added it ? It'll ensure that anyone reviewing all commits don't get to make comments that are in fact useless as the error is already corrected

Arthur-Milchior reviewed (commented):
I didn't review everything. I'd like you to clean commits and avoid adding and deleting things, so that I don't get to review things that will be removed.

I made suggestion of smaller changes that can be put in PR, that will be simpler to merge and will get this commit simpler to review and the code cleaner

bhatushar commented:
@Arthur-Milchior there are a lot of "undo" commits in this PR. I got a review from @david-allison-1 which required undoing a lot of changes I made.

I tried covering all major changes in https://github.com/ankidroid/Anki-Android/pull/8285#issuecomment-808295666. 

This PR has gotten too big with a lot of unnecessary commits. How can I fix this?

bhatushar commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -777,14 +766,53 @@ public boolean onKeyDown(int keyCode, KeyEvent event) {
+            try {
+                // Save changes to DB
+                mCol.getDecks().save(mOptions);
+                mCol.getDecks().save(mDeck);
```
This is something that can be done. I've mentioned something similar in the proposals section of https://github.com/ankidroid/Anki-Android/pull/8285#issuecomment-808295666. 

I'll implement this feature once the other changes are verified by a reviewer.

bhatushar commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -777,14 +766,53 @@ public boolean onKeyDown(int keyCode, KeyEvent event) {
+                setResult(RESULT_OK);
+                finish();
+                ActivityTransitionAnimation.slide(this, FADE);
+            } catch (RuntimeException e) {
```
That part of the code was there before I started working on it, so I just left it as is. If you think there's a better-suited exception for this, let me know.

david-allison-1 commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -184,6 +184,7 @@
 
     <!-- Deck Options -->
     <string name="deck_options_corrupt">Failed to process deck options: %s</string>
+    <string name="deck_options_group_changes" comment="XXX is dynamically replaced with options group name">Do you want to save changes for \'XXX\'?</string>
```
Arthur is correct, you should use '%s', we only use `XXX` in preferences due to the fact that we can't format the preference strings traditionally

david-allison-1 commented on AnkiDroid/src/main/res/menu/deck_options.xml:
```diff
@@ -0,0 +1,10 @@
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:ankidroid="http://schemas.android.com/apk/res-auto">
+
+    <item
```
Awesome!!!

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -198,139 +441,185 @@ public boolean commit() {
+                                Deck originalDeck = getDeckReference();
+                                DeckConfig newOptions = mCol.getDecks().getConf(newConfId);
+
+                                if (mPreferenceChanged) {
```
Can we extract this to a method?

bhatushar replied:
It's doable but I don't think it'll be much helpful. This section is only executed here. Rearranging cases might be better for readability.

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -361,43 +650,25 @@ public boolean commit() {
                             case "confSetSubdecks":
                                 if ((Boolean) value) {
-                                    TaskManager.launchCollectionTask(new CollectionTask.ConfSetSubdecks(mDeck, mOptions), confChangeHandler());
+                                    TaskManager.launchCollectionTask(new CollectionTask.ConfSetSubdecks(mDeckCopy, mOptionsCopy), confChangeHandler());
```
might want to alert that this will save

bhatushar replied:
It is part of Group Management. Please refer to the proposals section of https://github.com/ankidroid/Anki-Android/pull/8285#issuecomment-808295666. This can be resolved there but I think we should make a different PR for that.

Also, is it okay if I contact you via private message concerning some queries related to this PR?

david-allison-1 commented on AnkiDroid/src/main/res/values/03-dialogs.xml:
```diff
@@ -184,7 +184,7 @@
     <!-- Deck Options -->
     <string name="deck_options_corrupt">Failed to process deck options: %s</string>
-    <string name="deck_options_group_changes" comment="XXX is dynamically replaced with options group name">Do you want to save changes for \'XXX\'?</string>
+    <string name="deck_options_group_changes" comment="%s is dynamically replaced with options group name">Do you want to save changes for \'%s\'?</string>
```
```suggestion
    <string name="deck_options_group_changes">Do you want to save changes for \'%s\'?</string>
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -554,15 +554,14 @@ public boolean commit() {
                                 if (mPreferenceChanged) {
                                     // Preferences are changed, ask whether to save/discard before proceeding
                                     String confirmationText = getResources().getString(R.string.deck_options_group_changes);
```
```java
String title = mOptionsCopy.has("name") ? mOptionsCopy.getString("name") : "???";
String confirmationText = getResources().getString(R.string.deck_options_group_changes, title);
```

bhatushar replied:
Should I just amend these changes in the last commit?

david-allison-1 replied:
Basically: from our perspective, we have two choices when it comes to merging a PR:

* Rebase - copy each commit onto the head of master
* Squash - combine all commits into one, and add it to the head of master

For large changes, a rebase is better, as it lets each commit have its own description, and helps a developer break the change into logical steps.

But, a rebase means that each commit needs to be clean, and a piece of work on its own (no reverts, no TODOs).

Realistically there's 29 commits here, and it'll be a bit of work if you want to get it to a stage where we rebase, rather than squash.

In my opinion (but do check with @Arthur-Milchior), spending effort on cleaning a single commit is mostly wasted effort unless it's impossible for us to review, or it's planned to be able to get us to a point where we'll rebase the PR. 

Choice is yours

bhatushar replied:
Now that all major issues are resolved, I have a pretty clear idea of which commits are important and in which order they should be. 

But instead of rebase, I can do a soft reset and make new, cleaner commits. Idk if that's a good practice.

I'll wait for @Arthur-Milchior response.

Arthur-Milchior replied:
I always review code commit by commit. I find it really helpful to understand what is going on, and check each atomic change by itself. Instead of doing a big review, I got to make a lot of small easier review. 

When I got to undo a change or make a correction in it, I can just do an interactive rebase in order to rewrite history, to find the commit with the error, correct the error in it, and then submit it again with the correct code. That is what I would also expect of other people when I ask for correction in a commit.

Even if I were to agree that looking at the whole PR at once was acceptable (I could do it, it's big so it's not confortable, but it can be done), I still believe that the skill of doing interactive rebase, correcting history, and getting back to a simple to review PR will be useful to contribute in git-based project. 
However, I must also admit that after 29 commits, even I would have trouble figuring out what to correct where and I would probably just create a new branch. I'd look at the current branch to see what is the end goal and create atomic commits from it

bhatushar replied:
@Arthur-Milchior if it were small reverts, I would've done an interactive rebase. But after the review, I pretty much undid everything. I admit, it's far from my best work. 😅

If I reset this branch to the initial state and make new commits (keeping the final state in mind) will that be helpful to you for reviewing?

Arthur-Milchior replied:
Personally, I would have:
* created a new branch NB.
* Created the sequence of commits on NB.
* Renamed NB to 8180-deck-option-save
* force pushed

This would ensure that:
* you can keep a copy of current branch available until you are actually ready to lose all of its content
* put the new sequence on commit in the same PR for the sake of keeping the discussion history

Arthur-Milchior replied:
But just creating a new PR would be acceptable to me. Really, what I want is to be able to read a sequence of commit and think that each commit makes sens individually

david-allison-1 reviewed (changes requested):
Tiny changes

Just to note: from a style POV, we always want braces after an `if`

bhatushar commented:
As per @Arthur-Milchior's suggestion, I created a new series of commits that should be easier to review. The relevant commits start from e80bfee.

bhatushar commented:
@david-allison-1 @Arthur-Milchior 

Is anyone reviewing this PR?

david-allison-1 reviewed (changes requested):
Changes are not persisted if the user changes app

* Make a change to "new"
* Switch to GMail
* Switch Back
* Changes are reverted

bhatushar commented:
@david-allison-1 that could be a problem with the workflow implementation.

At the moment, changes are made to a local copy of data which is set inside `onCreate()`. I believe, when we switch back to the app, that method is called again resetting the changes.

The obvious fix is to apply changes to original data, instead of a copy. But that leads to the problem of reverting the changes programmatically.

I can't think of any other solution to this at the moment.

david-allison-1 commented:
> @david-allison-1 that could be a problem with the workflow implementation.
> 
> At the moment, changes are made to a local copy of data which is set inside `onCreate()`. I believe, when we switch back to the app, that method is called again resetting the changes.
> 
> The obvious fix is to apply changes to original data, instead of a copy. But that leads to the problem of reverting the changes programmatically.
> 
> I can't think of any other solution to this at the moment.

This is typically handled via `onSaveInstanceState` - we have a few examples in the codebase already

bhatushar commented:
@david-allison-1 I'll look into it tomorrow, it's already late at night here.

david-allison-1 commented:
> @david-allison-1 I'll look into it tomorrow, it's already late at night here.

Life comes first, we'll still be here, don't stress yourself

bhatushar commented:
> Changes are not persisted if the user changes app
> 
> * Make a change to "new"
> * Switch to GMail
> * Switch Back
> * Changes are reverted

I'm unable to reproduce it. Here's what I did:
* Made changes to New cards order, New cards/day, Graduation interval, Easy interval.
* Switched to different apps (Gmail, Chrome, Google) and browsed for a while.
* Returned to AnkiDroid.
* Changes were still there.

@david-allison-1 Maybe the device you tested on has set to clear cache as soon as you leave an app?

david-allison-1 commented:
Sorry! You're right. 

You need to enable "don't keep activities". Didn't realise I had it enabled

bhatushar commented:
So... do you want to persist temporary changes with "Don't keep activities" enabled?

david-allison-1 commented:
> So... do you want to persist temporary changes with "Don't keep activities" enabled?

It's best to persist via activity state and restore them when the activity is recreated, rather than persisting the changes to the database.

mikehardy commented:
Yep: all instance variables should go into the Bundle onSaveInstance state and when onCreate is called with the Bundle the Bundle should be examined for the same variables and re-initialized from the Bundle if they exist in there, as a rule. That's "the Android Activity Lifecycle" and is separate from user activity (although a user rotating their phone will trigger an Actviity lifecycle)

Database and prefs changes should only be made in response to actual user activity during normal execution

bhatushar commented:
@david-allison-1 @mikehardy Sorry for taking so long to get back to you. I got occupied with something else.

Anyway, I'm stuck on a problem. I'm saving the state data like Deck, Options, Changes, etc. as Mike suggested. When activity is resumed, all changes are loaded correctly. But when I'm trying to save a change in reminder time, the notification is not working.
The setting itself is saved. If I set the reminder to 11:24, it still shows 11:24 when I reopen the settings but there are no notifications. 

Any idea what could be causing it?

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -88,6 +88,43 @@
+
+    /**
+     * Returns a reference to the deck data.
+     * Uses {@link #mCol} to get deck data and if it's uninitialized, {@link NullPointerException} is thrown.
```
mCol is almost always initialized. It would be exceptional for it not to be.
If there is any reason to expect it not to be initialized here, it should be explained.
Otherwise, this case can be omitted from the comment

On the other hand, it would be helpful to get information in the comment about which deck data you get. I.e. deck from bundle, or current

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -88,6 +88,43 @@
+     *
+     * @return Reference to deck data
+     */
+    private Deck getDeckReference() throws NullPointerException {
```
Were you by any luck a C++ developer (or a rust one. Or any other low level language where actually using reference is standard practice)
FYI there is a class [Reference](https://docs.oracle.com/javase/7/docs/api/java/lang/ref/Reference.html) in java, and I think the only case where we use it is when we need weakReferences.
Technically, you're returning a pointer to an object on the heap, not a reference. But that's not something we usually need to consider (and I'm REALLY happy about it.) So in general, we just consider it to be a "Deck", and that's it.

So unless you actually need the notion of "reference" for some reason I can't figure out, I'd prefer you to remove this word of getDeckReference and getOptionsReference

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -88,6 +88,43 @@
+     * @return Reference to deck data
+     */
+    private Deck getDeckReference() throws NullPointerException {
+        if (mCol == null) {
```
Is there any reason to expect it to occur?

This seems useless to me. Indeed if you don't do it and directly access mCol below, we'll also get a NullPointerException, and we will also get as much information as we have in your message.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -688,12 +686,9 @@ protected void onCreate(Bundle icicle) {
-        }
+        mDeckCopy = getDeckReference().deepClone();
+        // TODO: Replace with deepClone() when #8350 is merged
+        mOptionsCopy = new DeckConfig(getOptionsReference().toString());
```
I added a "deepClone" into DeckConfig, so you can use it directly.

`toString` is relatively quick, but parsing from string is slow, it should be avoided

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -80,9 +80,9 @@
-
-    private DeckConfig mOptions;
-    private Deck mDeck;
+    // Copy of options and deck data is stored in the respective object
```
I am sorry but I do not understand this sentence.

Plus, it would be nice to have it in javadoc, since you document the meaning of a variable.
```java
/**
Current state of the deck configuration as displayed to the user. Potentially distinct from the state in the collection. Change in this object will be discarded unless they are saved in the deck manager
 */
```

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -688,12 +686,9 @@ protected void onCreate(Bundle icicle) {
-        } else {
-            mDeck = mCol.getDecks().current();
-        }
+        mDeckCopy = getDeckReference().deepClone();
```
I think that it would be nice if using `getDeckReference` and `getOptionsReference` could be moved to the first commit. It is simpler to review if the place the code is created is the same the place the code is introduced. Especially since it seems to mostly consist in moving the code from a place to another

Arthur-Milchior reviewed (commented):
Sorry for the time taken to answer. As already mentionned, big PR are hard to review, even if it seems a nice PR. And clearly, it's really really easier this time than on my first review, thank you VERY MUCH for the work you put in improving our work

> EDIT: I don't know if this is a good solution but I can try doing a soft reset and start making incremental commits, and force push to this branch.

Personally, when PR starts to become complex, this is what I do. But that's really exceptional, most of the time, I can juste use "interactive rebase" and get a proper commit history without needing to redo things from scratch.

In this cases, the first two commits are mostly good (I put a few, easy to fix, comments.). I consider that commits one and two are worth of their own PR. The reason are that:
* they improve code quality by themselves
* second commit is big, so it risks to take a lot of conflict. Merging it quickly ensure that we won't get conflict often
* it will be easier to concentrate on the next commits once the first one are in the codebase, since we know that we won't have to make more correction to the first commits

In any case, you'll need to rebase over master as there are conflicts.

In case you don't know, you can "cherry pick" a commit, so that you apply it directly. E.g. you can create a branch from master, then cherry pick your first commit, and then create a PR. It will ensure you'll have a PR that contains just your first commit and that we can review very quickly.

Hopefully, next review will come quicker. If PR are small, it's really really easier, and also there is less activity now that I've read all GSoC proposal we got

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -255,12 +262,12 @@ public boolean commit() {
+                                    mChanges.remove(key);
                                 }
-                                mOptionsCopy.getJSONObject("new").put("order", Integer.parseInt((String) value));
+                                mOptionsCopy.getJSONObject("new").put("order", newValue);
```
I find that `newValue` is really not descriptive. I'd clearly prefer something such as "newOrder"

I know you didn't choose the name.

I just submitted #8672 to improve it. If it gets merged, you'll get a merge conflict. You can also directly correct it yourself if you prefer

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java:
```diff
@@ -171,6 +228,206 @@ private void cacheValues() {
+         * Writes the changes made to {@link #mOptionsCopy} and {@link #mDeckCopy} to actual options and deck data.
+         * After writing the changes, they're committed to the database.
+         */
+        public void writeChanges() {
```
I don't understand why we need all of this. It seems duplicate work. Instead, you can just take the temporary `mOptionsCopy` and save it in the collection. That will replace the current value in the collection by the saved one. Or am I missing something?

Arthur-Milchior commented:
> EDIT: Sometimes the notification pops up minutes after the reminder time. This behavior is almost random.


How did you test that?

I'd note that most of the time, notification breaks just after a new install of ankidroid. In the sens that it will remark that it didn't notify in the last 24 hours and just notify anyway, even if it's not relevant. Really not practical when you install the app multiple time a day to check master. Might be related to problem you'd have here
